src/lib.rs

```diff
@@ -896,6 +1001,36 @@ impl<T> StableVec<T> {
         }
     }
 
+    /// Retains only the elements with indices specified by the given predicate.
+    ///
+    /// Each element with index `i` for which `should_be_kept(i)` returns `false` is
+    /// removed from the stable vector.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use stable_vec::StableVec;
+    /// let mut sv = StableVec::new();
+    /// sv.push(1);
+    /// let two = sv.push(2);
+    /// sv.push(3);
+    /// sv.retain_indices(|i| i == two);
+    ///
+    /// assert_eq!(sv, &[2] as &[_]);
+    /// ```
+    pub fn retain_indices<P>(&mut self, mut should_be_kept: P)
+    where
+        P: FnMut(usize) -> bool,
+    {
+        let mut pos = 0;
+
+        while let Some(idx) = next_valid_index(&mut pos, &self.deleted) {
+            if !should_be_kept(idx) {
+                self.remove(idx);
+            }
+        }
+    }
```

Comment:
Can you give me an example why one would need such a method?

Reply:
You mean pop_first()? Just seems logical to have it. I assume `Vec` doesn't have it because it would be inconsistent with `pop`, since `pop` is O(1) and `pop_first()` is O(n). But for `StableVec` both have the same complexity of O(n), so there's not reason to omit it.

Regarding the method names, I would leave `pop` for the sake of compatibility, add `remove_first` and `remove_last`, and reimplement `pop` as a call to `remove_first`

Reply:
I wrote retain_indices() for my compiler project. Nothing serious yet, just poking around IR in SSA form, control flow graphs and basic compiler stuff in general. SSA is all about variable definitions, and my IR has a `struct` for that, which contains a variable identifier (the left side) and a value index (the right side). A value can be a number, a string, a function reference, or an operation. So this program
```
let x = 3
f(1 + 2 + x, "a string")
```
without any optimizations becomes
```
v1 <- 3
v2 <- 1
v3 <- 2
v4 <- v2 + v3
v5 <- v4 + v1
v6 <- "a string"
v7 <- f
v8 <- call v7, v5, v6
```
Values are stored in a `ValueStorage` (sounds cheesy, I know): it has a `StableVec` for values and a `HashMap` for mapping values to indices. I apply a "constant folding" routine on values after I store them in the map (1 + 2 becomes just 3), then put the result in the vector. This is ok, because SSA values are supposed to be immutable, so if "v1 + v2" is found again, I can just use the existing value 3: `map[Add(v1, v2)]` is `0`, and `sv[0]` is `3` (it's already in the vector, since 3 was already assigned to `x`).

Then there is a dead code elimination pass. First, it removes the unused variables, then those that became unused after that and so on. Then it computes a set of used values and removes the rest from `ValueStorage` in one go by calling `retain` on the map and `retain_indices` on the vector (because values are only referenced in my IR by their indices in the vector).

Now, I cannot just iterate the vector (which provides `(index, &value)` tuples) and remove values at the same time, because that would require an allocation (`collect()`ing indices). And `retain` wants a predicate on values, not indices, and the whole point of using indices is to make things like dead code elimination cheap. Previously, I was using `Vec` and I had to remove the unused values from it (by moving contiguous holes to the end and then truncating the vector by calling `set_len`), it would invalidate the indices, so I had to remap all value usages to the new indices. Then I decided to use `StableVec` to get rid of remapping.

Sorry, I certainly need to improve my English and storytelling skills.

Reply:
I see, that makes sense! No worries, I'm fairly sure I understood your story and it's perfectly reasonable. Thanks for explaining.